File: javascript/space-age/space-age.js
//
// This is only a SKELETON file for the 'Space Age' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const age = (planet, ageInSeconds) => {
  const dayInSeconds = 24 * 60 * 60;       // 1 Earth day = 86400 seconds;
  const earthYear = dayInSeconds * 365.25; // 1 Earth year = 365.25 Earth days, or 31557600 seconds
  const earthAge = ageInSeconds / earthYear;
  var factor;

  switch (planet) {
    case "earth":
      factor = 1;
      break;

    case "mercury":
      factor = 0.2408467;
      break;

    case "venus":
      factor = 0.61519726;
      break;

    case "mars":
      factor = 1.8808158;
      break;

    case "jupiter":
      factor = 11.862615;
      break;

    case "saturn":
      factor = 29.447498;
      break;

    case "uranus":
      factor = 84.016846;
      break;

    case "neptune":
      factor = 164.79132;
      break;

    default:
      throw "Parâmetro inválido";
  }

  const age = earthAge / factor;

  return +(age.toFixed(2));
};

File: go/space-age/space_age.go
package space

// Planet enumeration
type Planet string

const (
	// Earth as a planet
	Earth Planet = "Earth"

	// Mercury as a planet
	Mercury = "Mercury"

	// Venus as a planet
	Venus = "Venus"

	// Mars as a planet
	Mars = "Mars"

	// Jupiter as a planet
	Jupiter = "Jupiter"

	// Saturn as a planet
	Saturn = "Saturn"

	// Uranus as a planet
	Uranus = "Uranus"

	// Neptune as a planet
	Neptune = "Neptune"
)

func (p Planet) float64() float64 {
	values := map[string]float64{"Earth": 1.0, "Mercury": 0.2408467, "Venus": 0.61519726, "Mars": 1.8808158, "Jupiter": 11.862615, "Saturn": 29.447498, "Uranus": 84.016846, "Neptune": 164.79132}

	x := string(p)

	if val, ok := values[x]; ok {
		return val
	}

	return -1
}

// year is 365.25 Earth days, or 31557600 seconds
const year = float64(31557600)

// Age function calculates the age in another planet years
func Age(seconds float64, planet Planet) float64 {
    factor := planet.float64()

    if factor < 0 {
    	return factor
    }

	return float64(seconds) / (year * factor)
}
